packages/test/snapshots/src/sanitize.ts

```diff
@@ -3,124 +3,692 @@
+    readonly defaultExcludedKeys = new Set<string>();
+    // Represents the keys used by merge-tree ops their "seg" property, where other
+    // keys represent user information
+    readonly mergeTreeExcludedKeys = new Set<string>();
+    // Map of user information to what it was replaced with.  Used to ensure the same
+    // data have the same replacements
+    readonly replacementMap = new Map<string, string>();
+
+    /**
+     * Validate that the provided message matches the provided schema.
+     * For a full scrub, warn and continue (scrubber should fully sanitize unexpected
+     * fields for ops), otherwise throw an error because we cannot be sure user
+     * information is being sufficiently sanitized.
+     */
+    objectMatchesSchema = (object: any, schema: any): boolean => {
+        const result =  schema === false ? falseResult : this.validator.validate(object, schema);
+        if (!result.valid) {
+            const errorMsg = `Bad msg fmt:\n${result.toString()}\n${JSON.stringify(object, undefined, 2)}`;
+
+            if (this.fullScrub || this.noBail) {
+                console.error(errorMsg);
+            } else {
+                throw new Error(errorMsg);
+            }
+        }
+        return result.valid;
+    };
+
+    readonly chunkProcessor = new ChunkedOpProcessor(this.objectMatchesSchema);
+
+    constructor(
+        readonly messages: ISequencedDocumentMessage[],
+        readonly fullScrub: boolean,
+        readonly noBail: boolean,
+    ) {
+        this.defaultExcludedKeys.add("type");
+        this.defaultExcludedKeys.add("id");
+        this.defaultExcludedKeys.add("pkg");
+        this.mergeTreeExcludedKeys.add("nodeType");
+    }
+
+    isFluidObjectKey(key: string): boolean {
+        return key === "type" || key === "id";
     }
 
-    const contents = messageContents.contents.content.contents;
-    if (contents.seg) {
-        if (contents.seg.text) {
-            contents.seg.text = replaceText(contents.seg.text);
-        } else if (typeof contents.seg === "string") {
-            contents.seg = replaceText(contents.seg);
+    /**
+     * Replace text with garbage.  FluidObject types are not replaced when not under
+     * full scrub mode.  Non-Generic type text is replaced consistently.
+     */
+    replaceText(input?: string, type: TextType = TextType.Generic): string {
+        if (input === undefined) {
+            return undefined;
+        }
+
+        if (type === TextType.Generic) {
+            return "xxxxx";
+        }
+
+        if (this.replacementMap.has(input)) {
+            return this.replacementMap.get(input);
+        }
+
+        let replacement;
+        switch (type) {
+            case TextType.Email:
+                // email values have a trailing "}" that is preserved here
+                replacement = `email_${uuid()}}`;
+                break;
+            case TextType.Name:
+                replacement = `name_${uuid()}`;
```

Comment:
i'm not sure there is much value in preserving email and name

Reply:
you should also match char length her

---
In reply to: [500444976](https://github.com/microsoft/FluidFramework/pull/3786#discussion_r500444976) [](ancestors = 500444976)